src/combinatorics/simple_graph/basic.lean

```diff
@@ -355,16 +355,107 @@ lemma complete_graph_is_regular [decidable_eq V] :
 by { intro v, simp }
 
 /--
-The minimum degree of all vertices
+The minimum degree of all vertices (and `0` if there are no vertices).
 -/
-def min_degree (G : simple_graph V) [nonempty V] [decidable_rel G.adj] : ℕ :=
-finset.min' (univ.image (λ (v : V), G.degree v)) (nonempty.image univ_nonempty _)
+def min_degree [decidable_rel G.adj] : ℕ :=
+option.get_or_else (univ.image (λ v, G.degree v)).min 0
```

Comment:
What's the advantage of this definition over `min'`? The lemmas below look like they should all be obvious statements about `min'` whose lemmas belong in the finset API.

Reply:
Those statement about `min'` are already in the finset API, except with additional assumptions. The advantage of this definition is that it doesn't require `V` to be nonempty - following the usual mathlib pattern of definitions having minimal assumptions, and having assumptions on the lemmas.

Reply:
Ah, I missed the removal of `nonempty V`. My point still stands though - none of your proofs are anything to do with graphs, and are essentially just stating properties of a general `finset.min_or_zero` function:

* `[nonempty V] → ∃ v, f v = min_or_zero f`
* `∀ v, min_or_zero f ≤ f v`
* `[nonempty V] → ∀ k, (∀ v, k ≤ f v) → k ≤ min_or_zero f`

I'd be tempted to either:
* Add an API like this to finset
* Work out why no one has needed this API yet, and use what the other code is already using.

Reply:
For instance, does this definition work?
```suggestion
finset.univ.sup (λ v, G.degree v)
```

Reply:
They're stating properties of your hypothetical `finset.min_or_zero` function, which applies to finsets and functions on them to the naturals - that is, it's just a composition of `finset.image`, `finset.min` and `option.get_or_else`. Are you suggesting we make an API for this particular composition rather than using the existing APIs for each of those?

Reply:
Yes, I'm suggesting we make that API, because what you have here is exactly that API anyway, just without generalizing over `f = (λ v, G.degree v)`.

Reply:
I don't think I understand your point here - what's special about this particular composition of three existing functions that means it deserves a special API? That is, we already have good APIs for the three individual functions, which is why there has been no need to make a special one for this combination - has something changed?

Reply:
I guess my objection comes in two parts:

* I'm not convinced the `0` default is actually a helpful API. An alternative would be to use `with_top ℕ` - the proofs in this PR remain essentially the same after adding two `with_top.coe_le_coe.mpr`s.
* If you did use `with_top ℕ`, you could redefine `min_degree` as `univ.inf (λ v, G.degree v)`, and then most of the proofs in this PR are actually missing API about `inf` on `linear_order`, which would be great to have directly on finset.

Reply:
The idea behind `with_top` was just to give the minimum of an empty set a well-behaved meaning - not to represent vertices of infinite degree.

I've added the missing `inf` lemma on `linear_order` in #6103.

Reply:
My reason for avoiding `with_top nat` is because in practice proving things in graph theory, it's much more helpful to return 0 than change the return type - again, this is the usual mathlib pattern, see eg `real.sqrt`.